### ISSUE TITLE
[Issue] GCR doesn't return a challenge when auth token has insufficient scope

### ISSUE THREAD

**Issue by jeromefroe** [regclient/regclient#149]
### Current Behavior

When trying to sync more than one image from Docker Hub to GCR any images after the first one will fail with "Empty challenge header" errors.

### Expected Behavior

All images should be sync'ed successfully.

### Steps To Reproduce

I'm encountering this issue with a private registry so I wasn't able to come up with a publicly reproducible example. If one has access to a private GCR registry then the following config file will reproduce the error:

```yaml
sync:
  - source: alpine
    target: gcr.io/<org>/alpine
    type: repository
    tags:
      allow:
        - "latest"
  - source: debian
    target: gcr.io/<org>/debian
    type: repository
    tags:
      allow:
        - "latest"
```

I've been running it with the following flags:

```console
regsync once -c regsync.yml -v debug
```

### Version

```text
{
  "VCSRef": "6a1a13c410f734f5e18a6032936bc6764814eae7",
  "VCSTag": "v0.3.10"
}
```

### Environment

- Running as binary or container: as a binary.
- Host platform: Darwin x86_64.
- Registry description: Docker Hub and GCR.

### Anything else

The cause of the issue appears to be that GCR is incorrectly failing to return a challenge when we try to sync the second image. Docker Hub correctly recognizes that the given auth token has insufficient scope and returns a 401 error with a challenge but GCR, while it does return a 401 error, fails to include a challenge. I've annotated the following logs to try to highlight the issue:

```text
# Initial HEAD request to Docker Hub for image A returns a 401 with auth challenge. 
DEBU[2022-02-06T15:47:21-05:00] http req                                      method=GET url="https://registry-1.docker.io/v2/<image A>/tags/list" withAuth=false
DEBU[2022-02-06T15:47:21-05:00] Auth request parsed                           challenge="[{bearer map[realm:https://auth.docker.io/token scope:repository:<image A>:pull service:registry.docker.io]}]"
# Subsequent requests to Docker Hub for image A with auth token are successful.
DEBU[2022-02-06T15:47:21-05:00] http req                                      method=GET url="https://registry-1.docker.io/v2/<image A>/tags/list" withAuth=true
DEBU[2022-02-06T15:47:22-05:00] http req                                      method=HEAD url="https://registry-1.docker.io/v2/<image A>/manifests/<tag 1>" withAuth=true
# Initial HEAD request to GCR for image A returns a 401 with auth challenge. 
DEBU[2022-02-06T15:47:22-05:00] http req                                      method=HEAD url="https://gcr.io/v2/<image A>/manifests/<tag 1>" withAuth=false
DEBU[2022-02-06T15:47:23-05:00] Auth request parsed                           challenge="[{bearer map[realm:https://gcr.io/v2/token scope:repository:<image A>:pull service:gcr.io]}]"
# Subsequent requests to GCR for image A with auth token are successful.
DEBU[2022-02-06T15:47:23-05:00] http req                                      method=HEAD url="https://gcr.io/v2/<image A>/manifests/<tag 1>" withAuth=true
DEBU[2022-02-06T15:47:23-05:00] Image matches                                 source="docker.io/<image A>:<tag 1>" target="gcr.io/<image A>:<tag 1>"
DEBU[2022-02-06T15:47:23-05:00] http req                                      method=HEAD url="https://registry-1.docker.io/v2/<image A>/manifests/<tag 2>" withAuth=true
DEBU[2022-02-06T15:47:23-05:00] http req                                      method=HEAD url="https://gcr.io/v2/<image A>/manifests/<tag 2>" withAuth=true
DEBU[2022-02-06T15:47:23-05:00] Image matches                                 source="docker.io/<image A>:<tag 2>" target="gcr.io/<image A>:<tag 2>"
# Request to Docker Hub for image B returns an "insufficient_scope" error.
DEBU[2022-02-06T15:47:24-05:00] http req                                      method=GET url="https://registry-1.docker.io/v2/<image B>/tags/list" withAuth=true
DEBU[2022-02-06T15:47:24-05:00] Auth request parsed                           challenge="[{bearer map[error:insufficient_scope realm:https://auth.docker.io/token scope:repository:<image B>:pull service:registry.docker.io]}]"
# Subsequent requests to Docker Hub for image B with updated auth token are successful.
DEBU[2022-02-06T15:47:24-05:00] http req                                      method=GET url="https://registry-1.docker.io/v2/<image B>/tags/list" withAuth=true
DEBU[2022-02-06T15:47:25-05:00] http req                                      method=HEAD url="https://registry-1.docker.io/v2/<image B>/manifests/<tag 3>" withAuth=true
# Request to GCR for image B returns an empty challenge instead of an "insufficient_scope" error.
DEBU[2022-02-06T15:47:25-05:00] http req                                      method=HEAD url="https://gcr.io/v2/<image B>/manifests/<tag 3>" withAuth=true
DEBU[2022-02-06T15:47:25-05:00] Auth request parsed                           challenge="[]"
WARN[2022-02-06T15:47:25-05:00] Failed to handle auth request                 Err="empty challenge header" URL="https://gcr.io/v2/<image B>/manifests/<tag 3>"

```

**Comment by sudo-bmitch** [comment]
Interesting, I wonder if I hit a regression with a bunch of my refactoring, or if GCR has changed something on their side. Previously I worked around their headers and response codes by manually invalidating my scope to force a new auth request rather than waiting for the 401. I'll get digging on this. Thanks for the report!

**Comment by sudo-bmitch** [comment]
Actually, I don't think I tested with more than one gcr repository, it was the push after a get/head that triggered some of my previous challenges. Either my AddScope isn't working (I'm second guessing it seeing the Hub auth responses), or it may be unhappy with me passing multiple scopes in a single request. More digging to do, but might take a bit since my schedule this week is looking busy.

**Comment by jeromefroe** [comment]
Thank you for the quick response @sudo-bmitch! One additional observation I made, if it helps, is that I was able to get things working by resetting the auth handler for a host when we encounter an empty challenge on the `v3.10` release (that's what I've been using). However, after opening the PR #150, I tried testing on the tip of `main` and started running into 403 unauthorized errors. I haven't had a chance to dig further, but my initial suspicion was that it may an issue with the images being processed concurrently.

**Comment by sudo-bmitch** [comment]
The 403 is probably from clearing the scope by resetting the auth on the second request. The first request still being run won't have it's score listed anymore. And instead of giving a 401 with the new scope needed, they go directly to 403. I've got an open issue with them on that behavior. In addition to that race condition, and the lag to wait for retries, I'm also hesitant to accept the PR because I'm guessing it won't work for coping between two gcr repositories. But it helps me know roughly where the issue is, so thanks for sending it over.

**Comment by sudo-bmitch** [comment]
@jeromefroe bad news and good news. Chatting with Google, me including multiple scopes in the auth request is very likely causing issues with GCR. And pulling that apart to do a separate auth per repo may take some work. For the good news, there are a bunch of options forward.

Option 1: You can run a separate regsync instance per GCR repository. Certainly not ideal, but would work.

Option 2: You can configure each repository as a separate pseudo registry entry in the yml, Something like `myproj.gcr.io` as the registry name in the host sections, and include `hostname: gcr.io` so we talk to the correct registry when using that name. Then on each sync line, you'd put `myproj.gcr.io` instead of `gcr.io` to give each their own auth. The downside is this won't work directly with docker's credentials and the helper, so you'd end up passing the user/pass logins through with an environment variable or secret file.

Option 3: Artifacts Registry works. It means you need to change your image names to `$region-docker.pkg.dev` instead of `gcr.io`, but also gives you all the other features they have with Artifacts Registry, and appears to be the push from Google to migrate. The downside is I believe it costs a bit more.

That's what I've got for now. Looking forward, there are two paths I'm thinking of. One is I can figure a decent way to sort that auth logic that doesn't break GCR, maybe as a configuration for the registry that triggers a separate auth per repo. I've got a rough plan for that code, just going to take a bit to write it up and test it.

The other thing I'm looking at is a bit of an overhaul of the credential handling, moving some of the calls to Docker APIs into my own code. I want to do more of a just-in-time logic that wouldn't get credentials for every registry on every command if we're only talking to one. As part of that change, I can make it easier to inject the credential helper logins for different pseudo registry names so they don't need to be passed as environment variables.

If you end up going with Artifacts Registry, let me know so I can bump this down in priority. Otherwise I can probably knock something out in a week.

**Comment by jeromefroe** [comment]
@sudo-bmitch thanks for the speedy response, it's much appreciated!! I'm actually not blocked on this (I've been using a build of the `v3.10` release with a change to reset the auth handler after an empty challenge error) so I would definitely advocate for whatever you think is the best approach even if it takes longer. Personally, I think I like your first approach of being able to get a separate auth per repo the best but I admittedly don't have enough context with the code base to know for certain.

**Comment by sudo-bmitch** [comment]
#155 just got merged that might apply here. I'm going to run some tests on my side later this week. Also have some AWS tests in my backlog. If there's still an issue, then next week I should be wrapping up my refactoring work and putting this on the front burner.

**Comment by sudo-bmitch** [comment]
Quick follow up that 155 didn't resolve the GCR issues, so no need to test that one. Doing some refactoring for a separate auth per repo (with a config setting since GCR is the only registry I've seen with this issue) is back to the top of my priority list now that my other refactoring changes are pushed out. I should have that included in a v0.4.0 release next week.

**Comment by sudo-bmitch** [comment]
This should be resolved in release 0.4.0 when using the `repoAuth: true` setting in the config. See [FAQ](https://github.com/regclient/regclient/tree/main/docs#faq) entry 4 for more details and let me know if you experience any other issues with this.

Speaking with Google on this, they should be able to support multiple pull scopes, but multiple push scopes will continue to be an issue that they are unlikely to fix with `gcr.io` (customer demand may change that). This setting may break the ability to perform cross-repo blob mounts, which means a push will be slower and use more bandwidth when copying between repositories on the same registry, so it won't be a default setting.

**Comment by jeromefroe** [comment]
Awesome! Thank you for all the help @sudo-bmitch!
